Add topic documentation on cloud billing concepts (#3773)
2i2c-org/infrastructure: docs/topic/billing/chargeable-resources.md

(topic:billing:resources:home)=
## Home directory

By default, we provide users with a persistent, POSIX compatible filesystem mounted
under `$HOME`. This persists across user server restarts, and is used to store
code (and sometimes data).

Because this has to persist regardless of whether the user is currently active or
not, cloud providers will charge us for it regardless of whether the user is actively
using it or not. Hence, choosing what cloud products we use here becomes crucial - if
we pick something that is a poor fit, storage costs can easily dwarf everything else.

The default from z2jh is to use [Dynamic Volume Provisioning](https://kubernetes.io/docs/concepts/storage/dynamic-provisioning/),
which sets up a *separate, dedicated* block storage device (think of it like a hard drive)
(([EBS](https://aws.amazon.com/ebs/) on AWS, [Persistent Disk](https://cloud.google.com/persistent-disk?hl=en) on GCP and [Disk Storage](https://azure.microsoft.com/en-us/products/storage/disks) on Azure))
*per user*. This has the following advantages:

1. No single user can take over the entire drive or use up extremely large amounts of

>    space, as the max size is enforced by the cloud provider.

Author: Great point, @GeorgianaElena. I've reworded and expanded this to clarify.

Each node needs some disk space that lives only as long as the node is up,
to store:

1. The basic underlying OS that runs the node itself (Ubuntu, [COS](https://cloud.google.com/container-optimized-os/docs) or similar)
2. Storage for any docker images that need to be pulled and used on the node
3. Temporary storage for any changes made to the *running containers* in a
   pod. For example, if a user runs `pip install` or `conda install` in a
   user server, that will install additional packages into the container
   *temporarily*. This will use up space in the node's disk as long as the
   container is running.
4. Temporary storage for logs
5. Any [emptyDir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir)
   volumes in use on the node. These volumes are deleted when the pod using
   them dies, so are temporary. But while they are in use, there needs to
   be 'some place' for this directory to live, and it is on the node's
   ephemeral disk
6. For *ephemeral hubs*, their home directories are also stored using the
   same mechanism as (3) - so they take up space on the ephemeral disk as
   well. Most hubs have a *persistent* home directory set up, detailed

>    in the [Home directories section](topic:billing:resources:home).

Reviewer: I kind of liked that persistent home storage was reference from here and I think it helps to make stronger connections between concepts. So I'd keep the last sentence as well, maybe prefixed by a but like: _"But most hubs have a ..."_

>    in the [Home directories section](topic:billing:resources:home).

Author: Yeah, I'd like to keep this section as well.

These disks can be fairly small - 100GB at the largest, but we can get away
with far smaller disks most of the time. The performance of these disks

> primarily matters *only* during initial docker image pull time - faster the

Author: ```suggestion
*primarily* matters during initial docker image pull time - faster the
```

disk, faster it is to pull large docker images. But it can be really
expensive to make the base disks super fast SSDs, so often a balanced
middling performance tier is more appropriate.

### Accelerators (GPU, etc)

Accelerators are a form of [co-processor](https://en.wikipedia.org/wiki/Coprocessor) -
dedicated hardware that is more limited in functionality than a general
purpose CPU, but much, *much* faster. [GPU](https://en.wikipedia.org/wiki/Graphics_processing_unit)
are the most common ones in use today, but there are also [TPU](https://cloud.google.com/tpu?hl=en)s,
[FGPA](https://en.wikipedia.org/wiki/Field-programmable_gate_array)s
available in the cloud. Depending on the cloud provider, these can either
be attached to *any* node type (GCP), or available only on a special set

> of node types & sizes (AWS & Azure).

Author: oooh, thanks for the catch, @consideRatio.

GPUs are the most commonly used with JupyterHubs, and are often the most
expensive as well! Leaving a GPU running for days accidentally is the
second easiest way to get a huge AWS bill (you'll meet NAT / network egress, the primary culprit, later in this document). So they are usually segregated
out into their own node pool that is spawned only for servers that require

> GPU use, and shut down after.

Author: What I was trying to communicate here is that you would never run non-GPU workloads on a GPU nodepool. I'll try to reword this to make that clearer.

> GPU use, and shut down after.

Author: I've reworded this to clarify, with a reference to kubernetes taints.

Depending on the source and destination of this transmission, **we may get charged for it**,
so it's important to have a good mental model of how network costs work when
thinking about cloud costs.

While the overall topic of network costs in the cloud is complex, thankfully
we can get away with a simplified model when operating just JupyterHubs.

### Network Boundaries

There are two primary aspects that determine how much a network
transmission costs:

1. The total amount of data transmitted
2. The **network boundaries** it crosses

(1) is fairly intuitive - the more data that is transmitted, the
more it costs. Usually, this is measured in 'GiB of data transmitted',
rather than smaller units.

(2) is determined by the **network boundaries** a data packet has to
cross from source to destination. Network boundaries can be categorized
in a million ways. But given we primarily want to model network costs,
we can consider a boundary to be crossed when:

1. The source and destination are in different zones of the same region of the cloud provider
2. The source and destination are in different regions of the same cloud provider

> 3. One of the source or destination is *outside* the cloud provider

Reviewer 2: ```suggestion
3. The source or destination is *outside* the cloud provider
```

> 3. One of the source or destination is *outside* the cloud provider

Author: ```suggestion
3. Either the source or destination is *outside* the cloud provider
```

   It is usually 1GiB in size, since that is the smallest disk we can get
   from many of these cloud providers. However, on *Azure*, we also store
   the hub logs on this disk for now - so the disk size is much larger.

   High performance is not a consideration here, so we usually try to use
   the cheapest (hence slowest disk) possible.

2. Prometheus disk

   We use [prometheus](https://prometheus.io/) to store time series data,
   both for operational (debugging) as well as reporting (usage metrics)
   purposes. We retain 1 year of data, and that requires some disk space.
   We try to not use the slowest disk type, but this doesn't require the
   fastest one either.

   Currently, most prometheus disks are bigger than they need to be. We
   can slim them down with effort if needed.

3. Grafana disk

   Grafana also uses a sqlite database the same way JupyterHub does. While
   most of the dashboards we use are [managed in an external repo](https://github.com/jupyterhub/grafana-dashboards),
   the database still holds user information as well as manually set up
   dashboards. This does not need to be fast, and we also try to have the

>    smallest possible disk here (1GiB)

Author: At least this section so far I'm trying to keep not particularly 2i2c specific to the extent possible, so I'm going to not include this one either. Hope that's ok.

   and this can get quite expensive
   (for reasons outlined in the ["Home Directory Storage" section](topic:billing:resources:home)).

Pricing is primarily dependent on the expected disk performance (measured as throughput and IOPS) as well as the disk size. While there's a lot of
options available, we'll list what we most commonly use in this table.

| Cloud Provider | Cost | Pricing Page |
| - | - | - |
| GCP | Standard (Slow) - $0.048 / GB / month, Balanced (Most common) - $0.12 / GB / month  | [link](https://cloud.google.com/compute/disks-image-pricing#disk) |
| AWS | gp2 (most common) - $0.1 / GB / month | [link](https://aws.amazon.com/ebs/pricing/) |
| Azure | ??? | [link](https://azure.microsoft.com/en-us/pricing/details/managed-disks/) |

## Kubernetes Cluster management

Cloud providers charge for maintaining the Kubernetes cluster itself,
separate from the nodes used to run various things in it. This is usually a *fixed* per hour cost per cluster.

1. [AWS EKS](https://aws.amazon.com/eks/pricing/)

   Costs $0.10 per hour flat.

2. [GCP GKE](https://cloud.google.com/kubernetes-engine/pricing#standard_edition)

>    Is a bit messier - it costs $0.10 per hour, but is structured in such a way
>    that one single-zone (non highly available) cluster is free per GCP project.
>    While this is a cost saving of about 75$ per month, in most cases we
>    use *regional* (multi-zonal) clusters that are not free. This makes
>    master upgrades much more painfree, and reduces overall amount of downtime
>    the hub has. So our default here is to pay the 75$ a month for the
>    regional cluster.

Reviewer 2: ```suggestion
   Costs $0.10 per hour flat.
   
   There is a discount of about 75$ per month allowing one *zonal* (single zone, non highly available) cluster per GCP billing account continuously for free. This discount is of little value to us and communities since since we favor the *regional* (multi-zone, highly available) clusters for less disruptive kubernetes control plane upgrades, and since we don't manage that many separate billing accounts.
```

These disks can be fairly small - 100GB at the largest, but we can get away
with far smaller disks most of the time. The performance of these disks
*primarily* matters during initial docker image pull time - faster the
disk, faster it is to pull large docker images. But it can be really
expensive to make the base disks super fast SSDs, so often a balanced
middling performance tier is more appropriate.

### Accelerators (GPU, etc)

Accelerators are a form of [co-processor](https://en.wikipedia.org/wiki/Coprocessor) -
dedicated hardware that is more limited in functionality than a general
purpose CPU, but much, *much* faster. [GPU](https://en.wikipedia.org/wiki/Graphics_processing_unit)
are the most common ones in use today, but there are also [TPU](https://cloud.google.com/tpu?hl=en)s,
[FGPA](https://en.wikipedia.org/wiki/Field-programmable_gate_array)s
available in the cloud. Accelerators can often only be attached to a subset of a cloud provider's available node types, node sizes, and cloud zones.

GPUs are the most commonly used with JupyterHubs, and are often the most
expensive as well! Leaving a GPU running for days accidentally is the

> second easiest way to get a huge AWS bill (you'll meet NAT / network egress, the primary culprit, later in this document). So they are usually segregated

Reviewer 3: > a huge AWS bill

Or any other cloud provider's bill as long as they offer GPUs (super minor, no need to fix, IMHO).

out into their own node pool that is _exclusively_ for use by servers that need
GPUs - a [kubernetes taint](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/)
is used to exclude all other user servers from those nodes. The cluster autoscaler
can stop the GPU node as soon as there are are actual GPU users, thus saving cost.

### Combining resource sizes when making a node

While you can often make 'custom' node sizes, *most commonly* you pick
one from a predefined list offered by the cloud provider ([GCP](https://cloud.google.com/compute/docs/machine-resource),
[AWS](https://aws.amazon.com/ec2/instance-types/),
[Azure](https://learn.microsoft.com/en-us/azure/virtual-machines/sizes)).
These machine types are primarily pre-defined sets of CPU (of a particular
generation, architecture and count), RAM and optionally
an accelerator (like GPU).

For the kinds of workloads we have, usually we pick something with a high
memory to CPU ratio, as notebooks (and dask workers) tend to be mostly
memory bound, not CPU bound.

### Autoscaling makes nodes temporary

> Nodes in our workloads are are *temporary* - they come and go with user

Reviewer 3: ```suggestion
Nodes in our workloads are *temporary* - they come and go with user
```

There are two primary aspects that determine how much a network
transmission costs:

1. The total amount of data transmitted
2. The **network boundaries** it crosses

(1) is fairly intuitive - the more data that is transmitted, the
more it costs. Usually, this is measured in 'GiB of data transmitted',
rather than smaller units.

(2) is determined by the **network boundaries** a data packet has to
cross from source to destination. Network boundaries can be categorized
in a million ways. But given we primarily want to model network costs,
we can consider a boundary to be crossed when:

1. The source and destination are in different zones of the same region of the cloud provider
2. The source and destination are in different regions of the same cloud provider
3. Either the source or destination is *outside* the cloud provider

If the source and destination are within the same zone, no network cost
boundaries are crossed and usually this costs no money. As a general rule,
when each boundary gets crossed, the cost of the data transmission
goes up. Within a cloud provider, transmission between regions is often
priced based on the distance between them - australia to

> the US is more expensive than canada to the US. Data transmission from within a cloud provider to outside is the most expensive, and is also often

Reviewer 3: ```suggestion
the US is more expensive than Canada to the US. Data transmission from within a cloud provider to outside is the most expensive, and is also often
```

how egress fees can hold your data 'hostage' in a cloud provider is finally causing
regulators to crack down, but only under limited circumstances (
[AWS](https://aws.amazon.com/blogs/aws/free-data-transfer-out-to-internet-when-moving-out-of-aws/),
[GCP](https://cloud.google.com/blog/products/networking/eliminating-data-transfer-fees-when-migrating-off-google-cloud)).

So when people are 'scared' of cloud costs, it's **most often** network
egress costs. It's very easy to make a bucket public, have a single other person
try to access it from another cloud provider, and get charged [a lot of money](https://2i2c.freshdesk.com/a/tickets/1105).

While scary, it's also fairly easy for us to watch out for.
A core use of JupyterHub is to put the compute near where the data is, so
*most likely* we are simply not going to see any significant egress fees
at all, as we are accessing data in the same region / zone. Object storage
access is the primary point of contention here for the kind of infrastructure
we manage, so as long as we are careful about that we should be ok. If you have data
in more than one region / zone, set up a hub in each region / zone you have data in -

> move the compute to where the data is, rather than the other way around.

Reviewer 3: Nice summary (and conceptual) last line!
